docs/design/inotify.md

```diff
@@ -0,0 +1,25 @@
+# Kata Containers support for `inotify`
+
+## Background on `inotify` usage
+
+A common pattern in Kubernetes is to watch for changes to files/directories passed in as `ConfigMaps`
+or `Secrets`. Sidecar's normally use `inotify` to watch for changes and then signal the primary container to reload
+the updated configuration. Kata Containers typically will pass these host files into the guest using `virtiofs`, which
+does not support `inotify`. This document describes how Kata Containers works around this limitation.
+
+### Detecting a `watchable` mount
+
+Kubernetes creates `secrets` and `ConfigMap` mounts at very specific locations on the host filesystem. For container mounts,
+the `Kata Containers` runtime will check the source of the mount to identify these special cases. For these use cases, only a single file
+or very few would typically need to be watched. To avoid excessive overheads in making a mount watchable,
+we enforce a limit of 8 files per mount. If a `secret` or `ConfigMap` mount contains more than 8 files, it will not be
+considered watchable. In the non-watchable scenario, updates to the mount continue to be propagated to the container workload,
+but these updates will not trigger an `inotify` event.
+
+### Presenting a `watchable` mount to the workload
+
+For mounts that are considered `watchable`, inside the guest, the `kata-agent` will poll the mount presented from
+the host through `virtiofs` and copy any changed files to a `tmpfs` mount that is presented to the container. In this way,
```

Comment:
Yeah, agreed.

In this case a bad actor would be in the namespace of the pod that is consuming the configmap, for example, and would be impacting just that single pod. Realistically I considered this to be someone shooting themselves in the foot, and worse case OOMing their containers.

Reply:
What do you think about adding some safe guards in the guest to limit tmpfs size to a sane limit (again, configmaps and certs/secrets are usually a single text file)?

If a file changes very often, that's okay. We only check for updates ever 2 sec (default today). If someone does this, we'll pick up the last change next cycle.

Reply:
I was thinking through this a bit more, and considering how to handle this safely, not failing silently, and not introducing too much extra complexity.

The initial implementation would check for the number of unique files in the configmap/secret at container creation time, and not enable polling (watcher) if the number is greater than 8. This was naive, though -- the number of files could increase/decrease over time (in particular for configmap, we'll actually add a file, then remove a file when the configmap is updated, so I still don't expect a large increase, but... expect the unexpected...). Similarly, right now we don't check the file(s) size, which as you point out could result in issues for the pod, since it'd be backed by tmpfs in the guest.

A suggestion: If we notice that the number of files or size of the files increase over a threshold for mount foo, we should:
- log an error
- remove the watcher / stop polling
- remove the contents at `MountPoint` of the watchable-bind storage object, and bind mount the storage objects `Source` over `MountPoint`

In this way, we stop polling, inotify stops working, but the contents of the mount are still available (and updated) in the container. This will match the functionality that exists today without the inotify handling.

WDYT?

Reply:
Updated docs to cover. this.